R/deduplicate.R

```diff
@@ -0,0 +1,206 @@
+
+  sql_file_info <- read_lines(orderly_file("database/file_info.sql"))
+  files <- DBI::dbGetQuery(con, sql_file_info)
+  files$path <- file.path(files$report,
+                          files$report_version,
+                          files$filename,
+                          fsep = "/")
+
+  paths <- file.path(config$root, "archive", files$path)
+  ## Information about the physical files, so we can work out which
+  ## files are already hardlinked
+  files <- cbind(
+    files,
+    fs::file_info(paths)[c("device_id", "inode", "hard_links")])
+  ## Information about if the files are unchanged
+  ## TODO: this step is slow for real files unfortunately
+  files$unchanged <- hash_files(paths, FALSE) == files$hash
+
+  ## Identify which files are duplicated
+  files$duplicated <-
+    files$hash %in% files$hash[duplicated(files$hash)]
+
+  ## Identify the first inode (in report id order)
+  i <- tapply(files$hash, files$hash)
+  files$inode_first <-
+    unname(tapply(files$inode, files$hash, function(x) x[[1L]]))[i]
+
+  ## Quick check:
+  stopifnot(all(vlapply(split(files, files$hash), function(x)
+    all(x$inode_first == x$inode[[1]]))))
+
+  ## Classify the files into different states
+  files$state <- rep("distinct", nrow(files))
+  files$state[files$duplicated] <- "duplicated"
+  files$state[files$duplicated & files$inode == files$inode_first] <- "linked"
+  files$state[files$duplicated & !duplicated(files$hash)] <- "distinct"
+
+  ## There is a corner case where if an input is an output (seen in
+  ## interactive in the demo) then we get into a right tangle.  So we
+  ## exclude the entry here for the artefact version of this.
+  files <- files[!duplicated(files$path), , drop = FALSE]
+  rownames(files) <- NULL
+
+  path_archive <- file.path(config$root, "archive")
+  all_files <- dir(path_archive, recursive = TRUE, all.files = TRUE,
+                   no.. = TRUE)
+  untracked <- setdiff(all_files, files$path)
+
+  exclude <- c("orderly_published.yml", ".orderly_archive_version", "README.md")
+  internal <- grepl("^orderly_run.*\\.(rds|yml)$", basename(untracked)) |
+    basename(untracked) %in% exclude
+
+  untracked <- data_frame(
+    path = untracked,
+    size = file_size(file.path(path_archive, untracked)),
+    internal = internal)
+
+  ret <- list(config = config,
+              files = files,
+              path = path_archive,
+              untracked = untracked)
+  class(ret) <- "orderly_deduplicate_info"
+  ret
+}
+
+
+orderly_deduplicate_prepare <- function(x) {
+  if (is_windows()) {
+    stop("Deduplication is not supported on windows")
+  }
+
+  assert_is(x, "orderly_deduplicate_info")
+
+  ## This is the only safe thing to do here:
+  if (length(unique(x$files$device_id)) > 1L) {
+    stop("Can't deduplicate as your orderly archive spans multiple devices")
+  }
+
+  ## Likely too strict as we can just skip these
+  if (!all(x$files$unchanged)) {
+    stop("Can't deduplicate files that have been modified")
```

Comment:
Should we print out which files have been modified?
(I got this error when I ran deduplicate on my machine and I don't know which files have been modified.)

Reply:
I've added a list here - it might get truncated for very large lists of files. Let me know how that goes on your machine